docs/proposals/20200506-conditions.md

```diff
@@ -0,0 +1,363 @@
+   // +optional
+   Message string `json:"message,omitempty" description:"human-readable message indicating details about last transition"`
+}
+
+// Conditions is the schema for the conditions portion of a Cluster API resource.
+type Conditions []Condition
+```
+
+Every Cluster API resource except data-only objects which aren't reconciled (i.e. `KubeadmConfigTemplate`)
+SHOULD have conditions field in the status struct. e.g.
+
+```golang
+// ClusterStatus represent the status for the Cluster object
+type ClusterStatus struct {
+    ...
+    // Conditions define a list of readiness conditions for the Cluster object
+    Conditions Conditions `json:"conditions,omitempty"
+} 
+```
+
+Each Cluster API resource SHOULD define its own set of condition types, e.g. 
+
+```golang
+// ConditionTypes for the cluster object 
+const (
+	ClusterInfrastructureReady ConditionType = "ClusterInfrastructureReady"
+	
+	...
+)
+```
+
+Condition types MUST have a consistent polarity (i.e. "True = good");
+
+Condition types MUST have one of the following suffix:
+
+- `Ready`, for resources which represent an ongoing status, like `ControlplaneReady` or `MachineDeploymentsReady`.
+- `Succeeded`, for resources which run-to-completion, e.g. `CreateVPCSucceeded`
+
+The `Severity` field MUST be set only when `Status=False` and it is designed to provide a standard classification
+of possible conditions failure `Reason`. 
+
+Please note that the combination of `Reason` and `Severity` gives different meaning to a condition failure
+allowing to detect when a long-running task is still ongoing:
+
+```
+ControlPlaneReady=False, Reason=ScalingUp, Severity=Info
+```
+
+In other cases, the combination of `Reason` and `Severity` allows to detect when a failure is due to a catastrophic
+error or to other events that are transient or can be eventually remediated by an user intervention
+
+```
+MachineReady=False, Reason=MachineNotHeathly, Severity=Error
+MachineReady=False, Reason=MachineUnderProbation, Severity=Warning
+AWSMachineReady=False, Reason=SSHKeyMissing, Severity=Warning
+```
+
+### Risks and Mitigations
+
+- Risk: This proposal is expected to be completed in three iterations, but in the initial stages this can make it
+  harder for the reviewers to figure out what will be the final result of this effort.
+- Mitigation: The incremental approach already embeds the possibility to make adjustments along the way. 
+
+- Risk: This proposal aims to be consistent with the target state of conditions in Kubernetes, but this is still
+  under definition (see [KEP](https://github.com/kubernetes/enhancements/pull/1624)).
+- Mitigation: Periodically re-evaluate this proposal vs the Kubernetes KEP. 
+
+- Risk: Cluster API presents some specific challenges that are not common to the core Kubernetes objects.
+- Mitigation: To allow a minimal set of carefully evaluated differences between Cluster API and Kubernetes
+  conditions.
```

Comment:
A lot of these points are repeated above, should we dedupe?

Reply:
Done